File: packages/kilo-vscode/src/agent-manager/worktree-diagnostics.ts
/**
 * Human-readable worktree health report.
 *
 * Exists because the previous answer to "why is Agent Manager empty?" was reading an output channel
 * full of per-poll failures and guessing. The report states what git and gh actually did, which
 * worktrees are unhealthy and why, and what is left on disk — in one place, copyable into an issue.
 *
 * The JetBrains plugin renders the same sections in the same order (see AgentManagerDiagnosticsAction)
 * so a report from either client can be read the same way.
 */

import { BUDGET } from "./command-budget"
import { GH } from "./gh"
import type { WorktreeHealthReport, WorktreeHealth } from "./worktree-reconcile"

export type ToolProbe = {
  /** Tool label as the report shows it, e.g. `git`. */
  name: string
  /** Version line, or undefined when the probe failed. */
  version?: string
  ms: number
  error?: string
}

export type DiagnosticsInput = {
  root: string
  worktreesDir: string
  probes: ToolProbe[]
  report: WorktreeHealthReport | undefined
  /** Worktrees currently skipped by the pollers, by id. */
  quarantined: string[]
  /** Row labels by worktree id, for a report that names things the way the UI does. */
  labels: Map<string, string>
}

const ORDER: WorktreeHealth[] = ["ok", "absent-restorable", "absent-gone", "unregistered", "unavailable"]

/** Render the report. Pure string building so it can be asserted in tests. */
export function diagnostics(input: DiagnosticsInput): string {
  const lines: string[] = []
  lines.push("Kilo Agent Manager — worktree health")
  lines.push(`repository: ${input.root}`)
  lines.push(`worktrees:  ${input.worktreesDir}`)
  lines.push("")

  lines.push("tools")
  for (const probe of input.probes) {
    const budget = probe.name === GH ? BUDGET.gh : BUDGET.probe
    const detail = probe.version ?? `FAILED — ${probe.error ?? "unknown error"}`
    lines.push(`  ${probe.name}: ${detail} (${probe.ms}ms, budget ${budget}ms)`)
  }
  lines.push("")

  const report = input.report
  if (!report) {
    lines.push("worktrees: health has not been determined yet")
    return lines.join("\n")
  }

  if (report.degraded) {
    // Worth stating plainly: every entry below reads `unavailable` for one reason, and no
    // self-healing ran, so the report must not be mistaken for "everything is broken".
    lines.push("NOTE: git could not list worktrees, so nothing was classified or repaired.")
    lines.push("")
  }

  const counts = new Map<WorktreeHealth, number>()
  for (const entry of report.entries) counts.set(entry.health, (counts.get(entry.health) ?? 0) + 1)
  lines.push("summary")
  for (const health of ORDER) lines.push(`  ${health}: ${counts.get(health) ?? 0}`)
  lines.push(`  orphan directories: ${report.orphans.length}`)
  lines.push(`  quarantined: ${input.quarantined.length}`)
  lines.push(`  pruned this pass: ${report.pruned}`)
  lines.push(`  state entries dropped: ${report.dropped.length}`)
  lines.push("")

  lines.push("worktrees")
  if (report.entries.length === 0) lines.push("  (none tracked)")
  for (const health of ORDER) {
    for (const entry of report.entries.filter((item) => item.health === health)) {
      const label = input.labels.get(entry.id) ?? entry.branch
      const flags = [
        `sessions=${entry.sessions}`,
        input.quarantined.includes(entry.id) ? "quarantined" : undefined,
      ].filter((flag) => flag !== undefined)
      lines.push(`  [${health}] ${label} — ${entry.path} (${flags.join(" ")})`)
    }
  }

  if (report.orphans.length > 0) {
    lines.push("")
    lines.push("orphan directories (nothing removes these automatically)")
    for (const orphan of report.orphans) lines.push(`  [${orphan.kind}] ${orphan.path}`)
  }
  return lines.join("\n")
}

/** Time a `--version` probe for one tool, so the report shows what the tools actually did. */
export async function probeTool(
  name: string,
  run: () => Promise<string>,
  now: () => number = Date.now,
): Promise<ToolProbe> {
  const started = now()
  try {
    return { name, version: (await run()).trim().split("\n")[0], ms: now() - started }
  } catch (error) {
    return { name, ms: now() - started, error: error instanceof Error ? error.message : String(error) }
  }
}

File: packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/agentManager/worktree/WorktreeDiagnostics.kt
package ai.kilocode.client.agentManager.worktree

import ai.kilocode.rpc.dto.GhAvailability
import ai.kilocode.rpc.dto.WorktreeDirtyDto
import ai.kilocode.rpc.dto.WorktreeDto
import ai.kilocode.rpc.dto.WorktreeStatsDto

/**
 * Renders the worktree-health report shown by the Advanced settings action.
 *
 * Sections and their order match the VS Code extension's diagnostics command, so a report from
 * either client reads the same way and can be compared directly in an issue. Pure string building:
 * the caller collects the data, this only formats it.
 */
internal object WorktreeDiagnostics {

    /** Everything the report needs, already fetched. */
    internal data class Input(
        val root: String,
        val gh: GhAvailability,
        val worktrees: List<WorktreeDto>,
        val orphans: List<String>,
        val stats: Map<String, WorktreeStatsDto>,
        val dirty: Map<String, WorktreeDirtyDto>,
    )

    fun render(input: Input): String {
        val lines = mutableListOf<String>()
        lines += "Kilo Agent Manager — worktree health"
        lines += "repository: ${input.root}"
        lines += ""

        lines += "tools"
        lines += "  git: ${if (input.gh == GhAvailability.GIT_MISSING) "NOT FOUND" else "ok"}"
        lines += "  gh:  ${ghLine(input.gh)}"
        lines += ""

        // "unavailable" means a poll could not measure the worktree — deliberately not folded into the
        // ok count, because that is exactly the conflation that made a failed poll look clean.
        val unavailable = input.worktrees.count { unavailable(input, it) }
        lines += "summary"
        lines += "  worktrees: ${input.worktrees.size}"
        lines += "  unavailable: $unavailable"
        lines += "  orphan directories: ${input.orphans.size}"
        lines += ""

        lines += "worktrees"
        if (input.worktrees.isEmpty()) lines += "  (none)"
        for (item in input.worktrees.sortedBy { !unavailable(input, it) }) {
            val key = normalizeWorktreePath(item.path)
            val state = if (unavailable(input, item)) "unavailable" else "ok"
            val reason = listOfNotNull(
                input.stats[key]?.reason?.takeIf { it.isNotBlank() },
                input.dirty[key]?.reason?.takeIf { it.isNotBlank() },
            ).firstOrNull()
            val detail = buildString {
                append("branch=${item.branch}")
                if (item.main) append(" main")
                if (item.locked) append(" locked")
                if (item.prunable) append(" prunable")
                if (reason != null) append(" reason=$reason")
            }
            lines += "  [$state] ${item.name} — ${item.path} ($detail)"
        }

        if (input.orphans.isNotEmpty()) {
            lines += ""
            lines += "orphan directories (nothing removes these automatically)"
            for (orphan in input.orphans) lines += "  $orphan"
        }
        return lines.joinToString("\n")
    }

    private fun unavailable(input: Input, item: WorktreeDto): Boolean {
        val key = normalizeWorktreePath(item.path)
        return input.stats[key]?.unavailable == true || input.dirty[key]?.unavailable == true
    }

    private fun ghLine(value: GhAvailability): String = when (value) {
        GhAvailability.OK -> "ok"
        GhAvailability.MISSING -> "NOT FOUND"
        GhAvailability.UNAUTH -> "not authorized"
        GhAvailability.RATE_LIMITED -> "rate limited"
        GhAvailability.TIMEOUT -> "did not answer within its budget"
        GhAvailability.GIT_MISSING -> "not probed (git missing)"
    }
}
